### ISSUE TITLE
Lack of isolation of where conditions

### ISSUE THREAD

**Issue by Godefroy** [typeorm/typeorm#1568]
**Issue type:** Bug report
**Database system/driver:** All SQL databases
**TypeORM version:** All

WHERE conditions generated by the query builder should all be isolated with parenthesis to avoid bugs due to operators priorities.

**Example:**

```ts
  const entries = await repository.createQueryBuilder()
    .where('field1 = :value OR field2 = :value', { value: 'foo' })
    .andWhere('field3 = :value', { value: 'bar' })
    .getMany();
```

Result conditions:
`WHERE field1 = 'foo' OR field2 = 'foo' AND field3 = 'bar'`

Expected conditions:
`WHERE (field1 = 'foo' OR field2 = 'foo') AND (field3 = 'bar')`

**Comment by pleerock** [comment]
its not a bug, its by design. If you want brackets isolation you can do following: 

```ts
        const users = await connection.createQueryBuilder(User, "user")
            .where("user.isAdmin = :isAdmin", { isAdmin: true })
            .orWhere(new Brackets(qb => {
                qb.where("user.firstName = :firstName1", { firstName1: "Timber" })
                    .andWhere("user.lastName = :lastName1", { lastName1: "Saw" });
            }))
            .orWhere(new Brackets(qb => {
                qb.where("user.firstName = :firstName2", { firstName2: "Alex" })
                    .andWhere("user.lastName = :lastName2", { lastName2: "Messer" });
            }))
            .getMany();
```

**Comment by pleerock** [comment]
Looks like the issue was resolved. Reopen if needed.

**Comment by RDeluxe** [comment]
I'm sorry, but how is this enormous and unreadable query could be by design ?

`.where('field1 = :value OR field2 = :value', { value: 'foo' }, brackets: true)` would be okayish. But the current implementation is far from optimal, readable and intuitive.

**Comment by dellert** [comment]
still has this bug

**Comment by ammanz** [comment]
```
.where(
        '(post.timePosted < :cursorTime OR (post.timePosted = :cursorTime AND post.id < :cursorId))',
        {
          cursorTime: cursorTime,
          cursorId: cursorId
        }
      );
```
I found that adding the required brackets straight into the where string works as expected. I find this more readable. The brackets syntax seems a bit wordy to me.

**Comment by fumieval** [comment]
We've been bitten by this issue today and __we almost leaked confidential information "by design"__!  Simply wrapping the expression with parentheses would have prevented this issue and I don't really see why not to do so. Does anyone __want__ `query.andWhere('A OR B')` to be interpreted as `(... AND A) OR B`?

**Comment by alper** [comment]
We'll have an ESlint rule to check for this nonsense.
